mixnet/util/src/lib.rs

```diff
@@ -0,0 +1,42 @@
+use std::{collections::HashMap, net::SocketAddr, sync::Arc};
+
+use parking_lot::Mutex;
+use tokio::net::TcpStream;
+
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct ConnectionPool {
+    pool: Arc<Mutex<HashMap<SocketAddr, Arc<tokio::sync::Mutex<TcpStream>>>>>,
+}
+
+impl ConnectionPool {
+    pub fn new(size: usize) -> Self {
+        Self {
+            pool: Arc::new(Mutex::new(HashMap::with_capacity(size))),
+        }
+    }
+
+    pub fn get(&self, addr: &SocketAddr) -> Option<Arc<tokio::sync::Mutex<TcpStream>>> {
+        self.pool.lock().get(addr).cloned()
+    }
+
+    #[allow(clippy::await_holding_lock)]
+    pub async fn get_or_init(
+        &self,
+        addr: &SocketAddr,
+    ) -> std::io::Result<Arc<tokio::sync::Mutex<TcpStream>>> {
+        let mut pool = self.pool.lock();
+        match pool.get(addr).cloned() {
+            Some(tcp) => Ok(tcp),
+            None => {
+                let tcp = Arc::new(tokio::sync::Mutex::new(TcpStream::connect(addr).await?));
```

Comment:
Ah ah sorry, I didn't notice that we have `connect().await` while acquiring lock.

After rethinking, it might not be very important to prevent temporary multiple connections for the same host because it will rarely happen probably. It's not easy to predict how critial this would be. Let's see and improve this gradually. Please feel free to correct me if I'm missing something.